fileio.c

```diff
@@ -487,17 +489,20 @@ int findfiles(char const *dir, void *data,
     while ((dent = readdir(dp))) {
 	if (dent->d_name[0] == '.')
 	    continue;
-	x_alloc(filename, strlen(dent->d_name) + 1);
-	strcpy(filename, dent->d_name);
-	r = (*filecallback)(filename, data);
-	if (r < 0)
-	    break;
-	else if (r > 0)
-	    filename = NULL;
+
+    filepath = getpathforfileindir(dir, dent->d_name);
+
+    if (!stat(filepath, &st) && S_ISDIR(st.st_mode)) {
+        if ((rres = findfiles(filepath, data, filecallback)) != TRUE)
+            return rres;
+    }
+    else {
+	    r = (*filecallback)(filepath, data);
+	    if (r < 0)
+    	    break;
+    }
```

Comment:
I'm not personally a fan of the mixed tab/space indentation style, but since the rest of the source is indented that way (4 spaces, 1 tab, 1 tab + 4 spaces, etc), it should be kept here as well instead of being inconsistent.